nni/nas/oneshot/pytorch/strategy.py

```diff
@@ -144,3 +147,92 @@ class RandomOneShot(OneShotStrategy):
 
     def __init__(self, **kwargs):
         super().__init__(RandomSamplingLightningModule, **kwargs)
+
+    def attach_model(self, base_model):
+        # This may conflict with the `self.run`` method. In the `run`, the base_model and evaluator are wrapped in `Model`. 
+        from nni.nas.oneshot.pytorch.base_lightning import traverse_and_mutate_submodules
+        _reason = 'The reason might be that you have used the wrong execution engine. Try to set engine to `oneshot` and try again.'
+
+        if not isinstance(base_model, nn.Module):
+            raise TypeError('Model is not a nn.Module. ' + _reason)
+        
+        extra_mixed_ops = getattr(base_model, "get_extra_mixed_oprations")() if hasattr(base_model, "get_extra_mixed_oprations") else []
+        self.mixed_ops_mappings = {op.bound_type:op for op in extra_mixed_ops+NATIVE_MIXED_OPERATIONS}
+
+        mutation_hooks = self.oneshot_kwargs.get("mutation_hooks", []) + RandomSamplingLightningModule.default_mutation_hooks(None)
+        module_list = traverse_and_mutate_submodules(base_model, mutation_hooks, RandomSamplingLightningModule.mutate_kwargs(None), topdown=True)
+        self.base_model = base_model
+        return module_list
+
+    def sub_state_dict(self, model):
```

Comment:
The fact that `sub_state_dict` needs a `model` as its argument is a bit strange, because it already has a `base_model`. It isn't actually doing anything to `model`, but only using it in a hacky way as a reference.

I think this hack was the only way in the past. But now, the cases are different. In the one-shot strategy, the modules like `PathSamplingLayer`, `PathSamplingInput`, `PathSamplingRepeat` can handle the `sub_state_dict` on their own. For example,

```python
class PathSamplingLayer:
  def sub_state_dict(arch: dict) -> dict:
    return self.paths[arch[self.label]].sub_state_dict(arch)

class MixedConv2d(MixedOperation):
  def sub_state_dict(arch: dict) -> dict:
    return {'weight': ..., 'bias': ...}
```

Then in `sub_state_dict` of random one-shot strategy, you traverse the `sub_state_dict` of all sub-modules and put them altogether. If `sub_state_dict` is not available, fallback to `state_dict`:

```python
class RandomOneShot(Strategy):
  def sub_state_dict(arch: dict) -> dict:
    ...
```

By doing so, I can think of the following benefits:

- You can get rid of all the crappy things like `STATE_DICT_PY_MAPPING`, which like I said, was in fact a work-around in the past,
- `sub_state_dict` doesn't depend on the instantiated model, but rather, it only uses an architecture dict, which is more light-weighted.
- You don't have to specifically handle mixed operations in this method, which decouples the strategy itself from module-specific logics.
- The interface of `slice_param` can be generalized to `state_dict` level.